### PR TITLE
GP11 : MAC functions fix/panic reason

### PR DESCRIPTION
**Pull request by cedric-chaumont-st-dev**
Signed-off-by: Cedric Chaumont <cedric.chaumont@st.com>

### CODE REVIEW

**Comment by cedric-chaumont-st-dev** [comment]
Updated.

File: lib/libutee/tee_api_operations.c [hunk 1]
```diff
@@ -860,75 +860,141 @@ TEE_Result TEE_CipherDoFinal(TEE_OperationHandle op,
...
+	switch (operation->info.algorithm) {
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_AES_CMAC:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
+		if (IV == NULL && IVLen != 0)
+			TEE_Panic(0);
```
Comment (pascal-brand-st-dev) [inline lib/libutee/tee_api_operations.c]: I would not panic as IV and IVLen are never used in the current algorithms

File: lib/libutee/tee_api_operations.c [hunk 2]
```diff
@@ -860,75 +860,141 @@ TEE_Result TEE_CipherDoFinal(TEE_OperationHandle op,
...
-	res = utee_hash_init(operation->state, IV, IVLen);
+
+	switch (operation->info.algorithm) {
+	case TEE_ALG_DES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_NOPAD:
+	case TEE_ALG_AES_CBC_MAC_PKCS5:
+	case TEE_ALG_AES_CMAC:
+	case TEE_ALG_DES_CBC_MAC_PKCS5:
+	case TEE_ALG_DES3_CBC_MAC_NOPAD:
+	case TEE_ALG_DES3_CBC_MAC_PKCS5:
```
Comment (pascal-brand-st-dev) [inline lib/libutee/tee_api_operations.c]: Missing HMAC algo.

File: lib/libutee/tee_api_operations.c [hunk 3]
```diff
@@ -860,75 +860,141 @@ TEE_Result TEE_CipherDoFinal(TEE_OperationHandle op,
...
+		if (IV == NULL && IVLen != 0)
+			TEE_Panic(0);
+		iv = IV;
+		ivlen = IVLen;
+		break;
+
+	default:
+		iv = NULL;
```
Comment (pascal-brand-st-dev) [inline lib/libutee/tee_api_operations.c]: I would panic as we should never reach default.

File: lib/libutee/tee_api_operations.c (hunk 1 above)
Comment (cedric-chaumont-st-dev) [inline lib/libutee/tee_api_operations.c]: Current algo listed use IV except CMAC ( I will remove it). Do you agree ?

File: lib/libutee/tee_api_operations.c (hunk 2 above)
Comment (cedric-chaumont-st-dev) [inline lib/libutee/tee_api_operations.c]: HAMC/CMAC algos do not use IV. Do you agree ?

File: lib/libutee/tee_api_operations.c (hunk 3 above)
Comment (cedric-chaumont-st-dev) [inline lib/libutee/tee_api_operations.c]: Even if currently not implemented, algos that could use IV should have coherent parameters. Do you agree ?

**Comment by cedric-chaumont-st-dev** [comment]
OK I will remove the check. IV=0 with CBC-MAC and keep the note.

**Comment by cedric-chaumont-st-dev** [comment]
Updated. Rebased.

File: lib/libutee/tee_api_operations.c [hunk 4]
```diff
@@ -860,75 +860,121 @@ TEE_Result TEE_CipherDoFinal(TEE_OperationHandle op,
...
 
 	if (operation == TEE_HANDLE_NULL)
 		TEE_Panic(0);
-	if (IV == NULL && IVLen != 0)
+	if (!operation->key1)
 		TEE_Panic(0);
 	if (operation->info.operationClass != TEE_OPERATION_MAC)
 		TEE_Panic(0);
-	res = utee_hash_init(operation->state, IV, IVLen);
+	/* Note : iv and ivlen are never used in current implementation */
```
Comment (pascal-brand-st-dev) [inline lib/libutee/tee_api_operations.c]: Can you add:

    This is why coherent values of IV and IVLen are not checked

**Comment by pascal-brand-st-dev** [comment]
Apart from a minor comment:

    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

**Comment by cedric-chaumont-st-dev** [comment]
Rebased.

**Comment by cedric-chaumont-st-dev** [comment]
Rebased.Squashed. Tested (STM, ARM Juno).
28701 subtests of which 0 failed
687 test cases of which 0 failed
0 test case was skipped
TEE test application done!

**Pull request merged**